<a href="https://colab.research.google.com/github/nikhilmamidipaka/Bank-Loan-Prediction-Dataset/blob/main/Bank_Loan_Prediction_Dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [93]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [94]:
# importing the useful libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [95]:
# Reading the application_data csv file using pandas
df_app = pd.read_csv("/content/drive/MyDrive/EDA PROJECT/Bank Loan Prediction Dataset/application_data.csv")
df_app


SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0           100002       1         Cash loans           M            N   
1           100003       0         Cash loans           F            N   
2           100004       0    Revolving loans           M            Y   
3           100006       0         Cash loans           F            N   
4           100007       0         Cash loans           M            N   
...            ...     ...                ...         ...          ...   
307506      456251       0         Cash loans           M            N   
307507      456252       0         Cash loans           F            N   
307508      456253       0         Cash loans           F            N   
307509      456254       1         Cash loans           F            N   
307510      456255       0         Cash loans           F            N   

       FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  \
0                    Y             0          202500.0    406597.5   
1                    N             0          270000.0   1293502.5   
2                    Y             0           67500.0    135000.0   
3                    Y             0          135000.0    312682.5   
4                    Y             0          121500.0    513000.0   
...                ...           ...               ...         ...   
307506               N             0          157500.0    254700.0   
307507               Y             0           72000.0    269550.0   
307508               Y             0          153000.0    677664.0   
307509               Y             0          171000.0    370107.0   
307510               N             0          157500.0    675000.0   

        AMT_ANNUITY  ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20  \
0           24700.5  ...                 0                0                0   
1           35698.5  ...                 0                0                0   
2            6750.0  ...                 0                0                0   
3           29686.5  ...                 0                0                0   
4           21865.5  ...                 0                0                0   
...             ...  ...               ...              ...              ...   
307506      27558.0  ...                 0                0                0   
307507      12001.5  ...                 0                0                0   
307508      29979.0  ...                 0                0                0   
307509      20205.0  ...                 0                0                0   
307510      49117.5  ...                 0                0                0   

       FLAG_DOCUMENT_21 AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
0                     0                        0.0                       0.0   
1                     0                        0.0                       0.0   
2                     0                        0.0                       0.0   
3                     0                        NaN                       NaN   
4                     0                        0.0                       0.0   
...                 ...                        ...                       ...   
307506                0                        NaN                       NaN   
307507                0                        NaN                       NaN   
307508                0                        1.0                       0.0   
307509                0                        0.0                       0.0   
307510                0                        0.0                       0.0   

        AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                              0.0                        0.0   
1                              0.0                        0.0   
2                              0.0                        0.0   
3                              NaN                        NaN   
4                              

In [96]:
# checking the columns description csv file using pandas
desc = pd.read_csv("/content/drive/MyDrive/EDA PROJECT/Bank Loan Prediction Dataset/columns_description.csv", encoding="latin")
desc

Unnamed: 0                     Table                        Row  \
0             1          application_data                 SK_ID_CURR   
1             2          application_data                     TARGET   
2             5          application_data         NAME_CONTRACT_TYPE   
3             6          application_data                CODE_GENDER   
4             7          application_data               FLAG_OWN_CAR   
..          ...                       ...                        ...   
155         209  previous_application.csv             DAYS_FIRST_DUE   
156         210  previous_application.csv  DAYS_LAST_DUE_1ST_VERSION   
157         211  previous_application.csv              DAYS_LAST_DUE   
158         212  previous_application.csv           DAYS_TERMINATION   
159         213  previous_application.csv  NFLAG_INSURED_ON_APPROVAL   

                                           Description  \
0                             ID of loan in our sample   
1    Target variable (1 - client with payment diffi...   
2          Identification if loan is cash or revolving   
3                                 Gender of the client   
4                        Flag if the client owns a car   
..                                                 ...   
155  Relative to application date of current applic...   
156  Relative to application date of current applic...   
157  Relative to application date of current applic...   
158  Relative to application date of current applic...   
159  Did the client requested insurance during the ...   

                                   Special  
0                                      NaN  
1                                      NaN  
2                                      NaN  
3                                      NaN  
4                                      NaN  
..                                     ...  
155  time only relative to the application  
156  time only relative to the application  
157  time only relative to the application  
158  time only relative to the application  
159                                    NaN  

[160 rows x 5 columns]

In [97]:
# displaying the first five columns
df_app.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21  \
0  ...                 0                0                0                0   
1  ...                 0                0                0                0   
2  ...                 0                0                0                0   
3  ...                 0                0                0                0   
4  ...                 0                0                0                0   

  AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        NaN                       NaN   
4                        0.0                       0.0   

   AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         NaN                        NaN   
4                         0.0                        0.0   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
0                        0.0                         1.0  
1                        0.0                         0.0  
2                        0.0                         0.0  
3                        NaN                         NaN  
4                        0.0                         0.0  

[5 rows x 122 columns]

In [98]:
# displaying the last five columns
df_app.tail()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
307506      456251       0         Cash loans           M            N   
307507      456252       0         Cash loans           F            N   
307508      456253       0         Cash loans           F            N   
307509      456254       1         Cash loans           F            N   
307510      456255       0         Cash loans           F            N   

       FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  \
307506               N             0          157500.0    254700.0   
307507               Y             0           72000.0    269550.0   
307508               Y             0          153000.0    677664.0   
307509               Y             0          171000.0    370107.0   
307510               N             0          157500.0    675000.0   

        AMT_ANNUITY  ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20  \
307506      27558.0  ...                 0                0                0   
307507      12001.5  ...                 0                0                0   
307508      29979.0  ...                 0                0                0   
307509      20205.0  ...                 0                0                0   
307510      49117.5  ...                 0                0                0   

       FLAG_DOCUMENT_21 AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
307506                0                        NaN                       NaN   
307507                0                        NaN                       NaN   
307508                0                        1.0                       0.0   
307509                0                        0.0                       0.0   
307510                0                        0.0                       0.0   

        AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
307506                         NaN                        NaN   
307507                         NaN                        NaN   
307508                         0.0                        1.0   
307509                         0.0                        0.0   
307510                         0.0                        2.0   

        AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
307506                        NaN                         NaN  
307507                        NaN                         NaN  
307508                        0.0                         1.0  
307509                        0.0                         0.0  
307510                        0.0                         1.0  

[5 rows x 122 columns]

In [99]:
# Finding the datatype of each column using info()
# As we have too many columns, we are using verbose=True
df_app.info(verbose= True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Data columns (total 122 columns):
 #    Column                        Dtype  
---   ------                        -----  
 0    SK_ID_CURR                    int64  
 1    TARGET                        int64  
 2    NAME_CONTRACT_TYPE            object 
 3    CODE_GENDER                   object 
 4    FLAG_OWN_CAR                  object 
 5    FLAG_OWN_REALTY               object 
 6    CNT_CHILDREN                  int64  
 7    AMT_INCOME_TOTAL              float64
 8    AMT_CREDIT                    float64
 9    AMT_ANNUITY                   float64
 10   AMT_GOODS_PRICE               float64
 11   NAME_TYPE_SUITE               object 
 12   NAME_INCOME_TYPE              object 
 13   NAME_EDUCATION_TYPE           object 
 14   NAME_FAMILY_STATUS            object 
 15   NAME_HOUSING_TYPE             object 
 16   REGION_POPULATION_RELATIVE    float64
 17   DAYS_BIRTH                    int64  
 18   DA

Concise summary of the dataframe, as we have too many columns, we are using the verbose = True mode.

The RangeIndex is from 0 to 307510 i.e. 307511 entries.

There are total 122 columns in the dataframe.

In [100]:
# describing the statistical measure of data
df_app.describe(include = 'all')

SK_ID_CURR         TARGET NAME_CONTRACT_TYPE CODE_GENDER  \
count   307511.000000  307511.000000             307511      307511   
unique            NaN            NaN                  2           3   
top               NaN            NaN         Cash loans           F   
freq              NaN            NaN             278232      202448   
mean    278180.518577       0.080729                NaN         NaN   
std     102790.175348       0.272419                NaN         NaN   
min     100002.000000       0.000000                NaN         NaN   
25%     189145.500000       0.000000                NaN         NaN   
50%     278202.000000       0.000000                NaN         NaN   
75%     367142.500000       0.000000                NaN         NaN   
max     456255.000000       1.000000                NaN         NaN   

       FLAG_OWN_CAR FLAG_OWN_REALTY   CNT_CHILDREN  AMT_INCOME_TOTAL  \
count        307511          307511  307511.000000      3.075110e+05   
unique            2               2            NaN               NaN   
top               N               Y            NaN               NaN   
freq         202924          213312            NaN               NaN   
mean            NaN             NaN       0.417052      1.687979e+05   
std             NaN             NaN       0.722121      2.371231e+05   
min             NaN             NaN       0.000000      2.565000e+04   
25%             NaN             NaN       0.000000      1.125000e+05   
50%             NaN             NaN       0.000000      1.471500e+05   
75%             NaN             NaN       1.000000      2.025000e+05   
max             NaN             NaN      19.000000      1.170000e+08   

          AMT_CREDIT    AMT_ANNUITY  ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19  \
count   3.075110e+05  307499.000000  ...     307511.000000    307511.000000   
unique           NaN            NaN  ...               NaN              NaN   
top              NaN            NaN  ...               NaN              NaN   
freq             NaN            NaN  ...               NaN              NaN   
mean    5.990260e+05   27108.573909  ...          0.008130         0.000595   
std     4.024908e+05   14493.737315  ...          0.089798         0.024387   
min     4.500000e+04    1615.500000  ...          0.000000         0.000000   
25%     2.700000e+05   16524.000000  ...          0.000000         0.000000   
50%     5.135310e+05   24903.000000  ...          0.000000         0.000000   
75%     8.086500e+05   34596.000000  ...          0.000000         0.000000   
max     4.050000e+06  258025.500000  ...          1.000000         1.000000   

       FLAG_DOCUMENT_20 FLAG_DOCUMENT_21 AMT_REQ_CREDIT_BUREAU_HOUR  \
count     307511.000000    307511.000000              265992.000000   
unique              NaN              NaN                        NaN   
top                 NaN              NaN                        NaN   
freq                NaN              NaN                        NaN   
mean           0.000507         0.000335                   0.006402   
std            0.022518         0.018299                   0.083849   
min            0.000000         0.000000                   0.000000   
25%            0.000000         0.000000                   0.000000   
50%            0.000000         0.000000                   0.000000   
75%            0.000000         0.000000                   0.000000   
max            1.000000         1.000000                   4.000000   

       AMT_REQ_CREDIT_BUREAU_DAY  AMT_REQ_CREDIT_BUREAU_WEEK  \
count              265992.000000               265992.000000   
unique                       NaN                         NaN   
top                          NaN                         NaN   
freq                         NaN                         NaN   
mean                    0.007000                    0.034362   
std                     0.110757                    0.204685   
min                     0.000000           

In [101]:
# Checking the number of rows and columns in dataset
df_app.shape

(307511, 122)

# Description of each column

In [104]:
desc_app1_data = desc.loc[desc.Table == 'application_data']

In [105]:
analyse_emptycol_df = pd.merge(left= emptycol_df,right=desc_appl_data,how='inner',on='Row')
analyse_emptycol_df[['Row','Description']]


NameError: ignored

## **Data Cleaning**

In [ ]:
# Checking the null values
df_app.isnull()

In [ ]:
# Checking the total number of null values in each feature in data frame
df_app.isnull().sum().values

In [ ]:
len(df_app.isnull())

In [ ]:
#Loans were paid on time having the Target variable is 0.
#Loans were not paid on time having the Target variable is 1.

df_app['TARGET'].value_counts().plot(kind="barh", figsize=(8,6))
plt.xlabel("Count", labelpad=12)
plt.ylabel("Target Variable", labelpad=12)
plt.title("Count of TARGET Variable per category", y=1.03);

In [ ]:
# Total percentage of 1s and 0s
100 * (df_app['TARGET'].value_counts())/len(df_app['TARGET'])

The above data is highly imbalanced and the ratio is 92:8.
    
Most of the loans were paid back on time having Target variable 0.


In [ ]:
emptycol= df_app.isnull().sum()
emptycol= emptycol[emptycol.values>(0.3 * len(emptycol))]

In [ ]:
emptycol_df = pd.DataFrame(emptycol)
emptycol_df = emptycol_df.reset_index()
emptycol_df = emptycol_df.rename(columns={"index": "Row", 0:"numcount"})
emptycol_df.head()

In [ ]:
missing = pd.DataFrame((df_app.isnull().sum())*100/df_app.shape[0]).reset_index()
plt.figure(figsize=(16,5))
ax = sns.pointplot('index',0,data=missing)
plt.xticks(rotation = 90, fontsize = 7)
plt.title("Percentage of Missing values")
plt.ylabel("PERCENTAGE")
plt.show()

Total columns having more than 30% of null values = 64

In [ ]:
# list out the missing values which are greater than 30%
emptycol = list(emptycol[emptycol.values > 0.3].index)

In [ ]:
emptycol.remove("AMT_GOODS_PRICE")
emptycol.remove("OWN_CAR_AGE")
emptycol.remove("OCCUPATION_TYPE")

In [ ]:
len(emptycol)

In [ ]:
df_app.drop(labels=emptycol, axis=1,inplace=True)
print(len(emptycol))

In [ ]:
# Checking the columns having null percentage
df_app.isnull().sum()/len(df_app)*100

In [ ]:
# Removing rows having null values greater than or equal to 30%

emptyrow = df_app.isnull().sum(axis=1)
emptyrow = list(emptyrow[emptyrow.values >= 0.3 * len(df_app)].index)
df_app.drop(labels = emptyrow, axis = 0, inplace = True)
print(len(emptyrow))

In [ ]:
# Filling missing values with median in column AMT_ANNUITY
values = df_app['AMT_ANNUITY'].median()
df_app.loc[df_app['AMT_ANNUITY'].isnull(),'AMT_ANNUITY']= values

In [ ]:
# Checking whether columns having null values
df_app.isnull().sum()

Now, all columns have been with zero null values

In [ ]:
df_app.AMT_CREDIT.plot.hist()

In [ ]:
plt.figure(figsize=[8,2])
sns.boxplot(df_app.AMT_CREDIT)
plt.show()

In [ ]:
sns.kdeplot(x='AMT_CREDIT',data = df_app, hue='CODE_GENDER')

In [ ]:
# Removing unwanted columns from the dataset

unwanted = ['FLAG_MOBIL', 'FLAG_EMP_PHONE', 'FLAG_WORK_PHONE', 'FLAG_CONT_MOBILE', 'FLAG_PHONE', 'FLAG_EMAIL', 'REGION_RATING_CLIENT', 'REGION_RATING_CLIENT_W_CITY', 'FLAG_EMAIL', 'CNT_FAM_MEMBERS', 'REGION_RATING_CLIENT', 'REGION_RATING_CLIENT', 'REGION_RATING_CLIENT_W_CITY', 'DAYS_LAST_PHONE_CHANGE', 'FLAG_DOCUMENT_2', 'FLAG_DOCUMENT_3', 'FLAG_DOCUMENT_4', 'FLAG_DOCUMENT_5', 'FLAG_DOCUMENT_6', 'FLAG_DOCUMENT_7', 'FLAG_DOCUMENT_8', 'FLAG_DOCUMENT_9', 'FLAG_DOCUMENT_10', 'FLAG_DOCUMENT_11', 'FLAG_DOCUMENT_12', 'FLAG_DOCUMENT_13', 'FLAG_DOCUMENT_14', 'FLAG_DOCUMENT_15', 'FLAG_DOCUMENT_16', 'FLAG_DOCUMENT_17', 'FLAG_DOCUMENT_18', 'FLAG_DOCUMENT_19', 'FLAG_DOCUMENT_20', 'FLAG_DOCUMENT_21']
df_app.drop(labels = unwanted, axis = 1, inplace = True)

In [ ]:
# Finding the categorical columns having the XNA(value is not available) values for Gender column

df_app[df_app['CODE_GENDER']=='XNA'].shape

In [ ]:
# Finding the categorical columns having the XNA(value is not available) values for Organization column

df_app[df_app['ORGANIZATION_TYPE']=='XNA'].shape

In [ ]:
# Describing the Gender column to check the number of females and males

df_app['CODE_GENDER'].value_counts()

In [ ]:
# Updating the column CODE_GENDER with 'F' for the dataset since female is having the majority and only 4 rows are having NA values

df_app.loc[df_app['CODE_GENDER'] == 'XNA', 'CODE_GENDER'] = 'F'
df_app['CODE_GENDER'].value_counts()

In [ ]:
# Describing the organization type column

df_app['ORGANIZATION_TYPE'].describe()

In [ ]:
# dropping the rows of total 55374 having 'XNA' values in the organization type column

data = df_app.drop(df_app.loc[df_app['ORGANIZATION_TYPE'] == 'XNA'].index)
data[data['ORGANIZATION_TYPE'] == 'XNA'].shape

## Categorical Analysis

In [ ]:
color = ['blue', 'orange', 'green', 'red', 'violet', 'pink']

df_app.CODE_GENDER.value_counts(normalize=True).plot.bar(color=color)
plt.show()
df_app.NAME_FAMILY_STATUS.value_counts(normalize=True).plot.bar(color=color)
plt.show()
df_app.NAME_INCOME_TYPE.value_counts(normalize=True).plot.bar(color=color)
plt.show()
df_app.OCCUPATION_TYPE.value_counts(normalize=True).plot.bar(color=color)
plt.show()



By just analysing single variables, we won't find much insights related to the defaulters, as here we will just have an idea which category of people are present in abundance, other than that, most of the insights are gathered in analysis of multiple features/variables with target variable.

In [ ]:
df_app.NAME_FAMILY_STATUS.value_counts(normalize=True).plot.pie()
plt.show()

In [ ]:
df_app.plot.scatter(x='SK_ID_CURR',y='CNT_CHILDREN')
plt.show()

## **Income Type vs Target Variable**

In [ ]:
def plot_features(feature,label_rotation = False,horizontal_layout = True):
  temp = df_app[feature].value_counts()
  df1_app = pd.DataFrame({feature: temp.index,'absolute values': temp.values})
  t1_percentage = df_app[[feature, 'TARGET']].groupby([feature],as_index=False).mean()
  t1_percentage.sort_values(by='TARGET', ascending=False, inplace=True)
  if(horizontal_layout):
        fig, (ax1, ax2) = plt.subplots(ncols=2, figsize=(12,6))
  else:
        fig, (ax1, ax2) = plt.subplots(nrows=2, figsize=(12,14))
  s = sns.barplot(ax=ax1, x = feature, y="absolute values",data=df1_app)
  if(label_rotation):
      s.set_xticklabels(s.get_xticklabels(),rotation=90)
    
  s = sns.barplot(ax=ax2, x = feature, y='TARGET', order=t1_percentage[feature], data=t1_percentage)
  if(label_rotation):
      s.set_xticklabels(s.get_xticklabels(),rotation=90)
  plt.ylabel('% of applications where loans were NOT repayed', fontsize=10)
  plt.tick_params(axis='both', which='major', labelsize=10)
  plt.subplots_adjust(wspace = .6)
  plt.show()
  

In [ ]:
plot_features('NAME_FAMILY_STATUS',True,True)

From the above barplots, I observed that most clients are married. Civil marriages and single people have higher chances of defaulting than others.

In [ ]:
plot_features('NAME_INCOME_TYPE',False, False)

From the above barplots, I observed that working class clients applies the most for loans, have a very low default rate, hence they are reliable.

Clients who are on Maternity leave or unemployed have very high defaults rates even though they are a minority compared to other income types.

Commercial associates, state servants and pensioners are fairly more reliable.